_posts/2017-11-24-upgrade-rails-from-4-0-to-4-1.markdown

```diff
@@ -0,0 +1,226 @@
+
+If you thought default scopes on models could be confusing, there's even another
+(un?)expected twist to it:
+
+```ruby
+default_scope { where active: true }
+scope :inactive, -> { where active: false }
+```
+
+```ruby
+User.all
+# SELECT "users".* FROM "users" WHERE "users"."active" = 'true'
+
+User.inactive
+# SELECT "users".* FROM "users" WHERE "users"."active" = 'false'
+```
+
+```ruby
+User.all
+# SELECT "users".* FROM "users" WHERE "users"."active" = 'true'
+
+User.inactive
+# SELECT "users".* FROM "users" WHERE "users"."active" = 'true' AND "users"."active" = 'false'
+```
+
+If you depended on this behavior, you will need to work around it using
+`unscoped`, `unscope` or the new `rewhere` method (https://github.com/rails/rails/commit/f950b2699f97749ef706c6939a84dfc85f0b05f2).
+
+(Friendly reminder: beware when using [default_scope](https://www.ombulabs.com/blog/ruby/rails/best-practices/why-using-default-scope-is-a-bad-idea.html))
+
+- No more mutable methods on ActiveRecord relations:
+
+`ActiveRecord::Relation` no longer has access to mutator methods like `#map!`,
+`#delete_if` or `#compact!`. If you need to use them, you will need to convert
+the `Relation` to an `Array` by calling `#to_a` first.
+
+Before:
+```ruby
+Project.where(title: 'Rails Upgrade').compact!
+```
+
+After:
+```ruby
+projects = Project.where(name: 'Rails Upgrade').to_a
+projects.compact!
+```
+
+- Implicit joins are removed:
+
+Before Rails 4.1, if you had this code:
+
+```ruby
+Post.includes(:comments).where("comments.title = 'foo'")
+```
+
+ActiveRecord 4.0 would know to join `posts` and `comments` in the executed SQL,
+and it would work just fine.
+
+However, Rails shouldn't have to be smart and parse your `where` with a regular
+expression to figure out what tables you want to join, since it leads to
+bugs (https://github.com/rails/rails/issues/9712).
+
+To fix this problem, you need to use an explicit join:
+
+```ruby
+Post.joins(:comments).where("comments.title = 'foo'")
+```
+
+UNLESS your intention was to actually eager load the post's comments.
+In that case, you can use the following syntax:
+
+```ruby
+Post.eager_load(:comments).where("comments.title = 'foo'")
+```
+
+Or:
+
+```ruby
+Post.includes(:comments).where("comments.title = 'foo'").references(:comments)
+```
+
+Both are equivalent and produce the same SQL.
+
+For a more in depth explanation of the differences between `joins`, `includes`,
+`references`, `eager_load` and even `preload`, [check out this post](http://blog.ifyouseewendy.com/blog/2015/11/11/preload-eager_load-includes-references-joins/).
+
+Finally, if you get this deprecation warning:
+
+```
+DEPRECATION WARNING: Implicit join references were removed with Rails 4.1.Make sure to remove this configuration because it does nothing.
+```
+
+Just remove `config.active_record.disable_implicit_join_references` from your
+config files.
+
+<h2 id="tests">5. Tests</h2>
+
+- CSRF protection now covers GET requests with JS responses:
```

Comment:
You can add a link to `CSRF protection` (http://guides.rubyonrails.org/security.html#cross-site-request-forgery-csrf)